### PR TITLE
Wrong call hierarchy computed for type in conflicting modules

### PR DESCRIPTION
**Pull request by trancexpress**
JavaSearchNameEnvironment adds module infos to each PackageFragmentRoot of a project, whether the root fragment is on the compile classpath or module path. However the module info is only valid for fragments that are on the module path.

As a result, a search (call hierarchy, references) can encounter non-existent compile errors such as module conflicts. The project itself compiles fine, but the search detects such conflicts due to the module infos attached to each PackageFragmentRoot.

This change adjusts JavaSearchNameEnvironment to not add module info for a PackageFragmentRoot which is not on the module path of the project.

Fixes: #675

## Author checklist

- [ ] I have thoroughly tested my changes
- [ ] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [ ] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by trancexpress** [comment]
@iloveeclipse WDYT?

I'm looking into adding a test next.

File: org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/JavaSearchNameEnvironment.java [hunk 1]
```diff
@@ -563,4 +573,18 @@ private static boolean isComplianceJava9OrHigher(IJavaProject javaProject) {
 	}
 	return CompilerOptions.versionToJdkLevel(javaProject.getOption(JavaCore.COMPILER_COMPLIANCE, true)) >= ClassFileConstants.JDK9;
 }
+
+private static boolean isOnModulePath(PackageFragmentRoot root) throws JavaModelException {
+	IClasspathEntry classpathEntry = root.getRawClasspathEntry();
+	IClasspathAttribute[] attributes = classpathEntry.getExtraAttributes();
```
Comment (iloveeclipse) [inline org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/JavaSearchNameEnvironment.java:579]: I see similar code in `org.eclipse.jdt.internal.core.ClasspathEntry.isModular()` and `org.eclipse.jdt.launching.JavaRuntime.isModule(IClasspathEntry, IJavaProject)`.

Looks like this should be just a default method on `IClasspathEntry`

**Comment by iloveeclipse** [comment]
Make sense, but see comment above regarding attribute checks.

File: org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/JavaSearchNameEnvironment.java (hunk 1 above)
Comment (trancexpress) [reply org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/JavaSearchNameEnvironment.java]: If we want to introduce new API, best to open an extra issue for that. Can you check the current change? Is that enough? If so, should I adjust also `JavaRuntime` (that seems to be in another plug-in) to use the new method? Though it might be accessing internal code then.

File: org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/JavaSearchNameEnvironment.java [hunk 2]
```diff
@@ -281,7 +282,16 @@ private ClasspathLocation mapToClassPathLocation(JavaModelManager manager, Packa
 	}
 	JavaProject javaProject = root.getJavaProject();
 	if (isComplianceJava9OrHigher(javaProject)) {
-		addModuleClassPathInfo(root, defaultModule, cp);
+		boolean isOnModulePath = true; // if an exception occurs, assume yes
+		try {
+		    IClasspathEntry classpathEntry = root.getRawClasspathEntry();
+			isOnModulePath = ClasspathEntry.isModular(classpathEntry);
```
Comment (iloveeclipse) [inline org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/JavaSearchNameEnvironment.java:288]: Something is wrong with indentation here

File: org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/JavaSearchNameEnvironment.java (hunk 1 above)
Comment (iloveeclipse) [reply org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/JavaSearchNameEnvironment.java]: Right, let split that in two changes.

**Comment by iloveeclipse** [comment]
@trancexpress : either I've messed something with the merge (test had merge issues) or the changes on master cause now fails. Could you please check with your local branch copy?

**Comment by iloveeclipse** [comment]
Actually few Java 9+ search tests have now failures. I assume they worked **without** module attributes set on classpath entries.

**Comment by trancexpress** [comment]
I'll check tomorrow.

  | Test Result (30 failures  / -1)org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_008org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_009org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_012org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_013org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_014org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_015org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_016org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_017org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_018org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_019org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_020org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_021org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_022org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_027org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_028org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_029org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_030org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_031org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_032org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_041org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_043org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_045org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug519980_001org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug519980_002org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug520477_001org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug519151_023org.eclipse.jdt.core.tests.model.JavaSearchBugs15Tests.test566062_001org.eclipse.jdt.core.tests.model.JavaSearchBugs15Tests.test566062_002org.eclipse.jdt.core.tests.model.JavaSearchBugs15Tests.test566062_003org.eclipse.jdt.core.tests.model.JavaSearchBugs15Tests.test566062_004
-- | --


[Test Result](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/) (30 failures / -1)

    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_008](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_008/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_009](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_009/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_012](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_012/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_013](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_013/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_014](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_014/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_015](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_015/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_016](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_016/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_017](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_017/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_018](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_018/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_019](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_019/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_020](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_020/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_021](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_021/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_022](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_022/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_027](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_027/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_028](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_028/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_029](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_029/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_030](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_030/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_031](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_031/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_032](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_032/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_041](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_041/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_043](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_043/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug501162_045](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug501162_045/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug519980_001](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug519980_001/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug519980_002](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug519980_002/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug520477_001](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug520477_001/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs9Tests.testBug519151_023](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs9Tests/testBug519151_023/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs15Tests.test566062_001](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs15Tests/test566062_001/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs15Tests.test566062_002](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs15Tests/test566062_002/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs15Tests.test566062_003](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs15Tests/test566062_003/)
    [org.eclipse.jdt.core.tests.model.JavaSearchBugs15Tests.test566062_004](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-679/9/testReport/junit/org.eclipse.jdt.core.tests.model/JavaSearchBugs15Tests/test566062_004/)

**Comment by trancexpress** [comment]
There is some weird classpath entry that is added by some of the test... E.g.:

```
<?xml version="1.0" encoding="UTF-8"?>
<classpath>
	<classpathentry kind="src" path="src"/>
	<classpathentry kind="var" path="JCL19_LIB">
		<attributes>
			<attribute name="module" value="true"/>
		</attributes>
	</classpathentry>
	<classpathentry kind="con" path="org.eclipse.jdt.MODULE_PATH"/>
	<classpathentry kind="src" path="/second"/>
	<classpathentry kind="output" path="bin"/>
</classpath>
```

Its maybe related to this bug, when `JavaCore.MODULE_PATH_CONTAINER_ID` was added: https://bugs.eclipse.org/bugs/show_bug.cgi?id=479610

https://github.com/eclipse-jdt/eclipse.jdt.core/commit/9b4c6fe06310162fcf691a59705683aa5c5ce814

```
	/**
	 * @since 3.14
	 */
	public static final String MODULE_PATH_CONTAINER_ID = "org.eclipse.jdt.MODULE_PATH"; //$NON-NLS-1$
```

The tests seem to add a container with this path:

```
addClasspathEntry(project2, JavaCore.newContainerEntry(new Path("org.eclipse.jdt.MODULE_PATH")));
```

I don't know what the significance of entry is, or how its supposed to indicate what is on the module path. But I think as soon as we see one, we should not make assumptions as to what is on the module path and what isn't. I've adjusted the change accordingly.

File: org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/JavaSearchNameEnvironment.java [hunk 3]
```diff
@@ -563,4 +573,15 @@ private static boolean isComplianceJava9OrHigher(IJavaProject javaProject) {
 	}
 	return CompilerOptions.versionToJdkLevel(javaProject.getOption(JavaCore.COMPILER_COMPLIANCE, true)) >= ClassFileConstants.JDK9;
 }
+
+private static boolean hasModulePathContainer(JavaProject javaProject) throws JavaModelException {
+	IClasspathEntry[] classpath = javaProject.getRawClasspath();
+	for (IClasspathEntry entry : classpath) {
+		IPath entryPath = entry.getPath();
+		if (entryPath != null && JavaCore.MODULE_PATH_CONTAINER_ID.equals(entryPath.toString())) {
```
Comment (iloveeclipse) [inline org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/JavaSearchNameEnvironment.java:581]: This sounds like a hack on top of another hack. Would be great to understand what that `MODULE_PATH_CONTAINER` is good for and if that should be better properly configured instead. I believe at the time it was introduced there were no "module" classpath attribute.

Comment (trancexpress) [reply org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/JavaSearchNameEnvironment.java:581]: And? Projects like that will exist, same as during the tests. Just because its a hack doesn't mean its not being used.

Comment (iloveeclipse) [reply org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/JavaSearchNameEnvironment.java:581]: Simeon, you didn't get it. I mean, this container looks like a hack for JDT tests, nothing more. It will be NOT exist in any code outside JDT.

Comment (trancexpress) [reply org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/JavaSearchNameEnvironment.java:581]: Hi @mpalat ,

can you clarify the purpose and usage of `JavaCore.MODULE_PATH_CONTAINER_ID`/`ModulePathContainer`?

Was this some hack to test the compiler module path with Eclipse projects? Can we expect that projects having a classpath entry as follows, are only used in tests and don't exist as actual projects in "real" workspaces?

```
<classpathentry kind="con" path="org.eclipse.jdt.MODULE_PATH"/>
```

If this was added for tests only, we can adjust the tests that have problems with the change (to have a proper module path).

Best regards and thanks,
Simeon

Comment (trancexpress) [reply org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/JavaSearchNameEnvironment.java:581]: I've prepared a 2nd branch with how tests would look without using `JavaCore.MODULE_PATH_CONTAINER_ID`:

https://github.com/trancexpress/eclipse.jdt.core/commit/404f642d03eb2f1e32013552867a24d595db176a

Its not a small change.

Comment (iloveeclipse) [reply org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/JavaSearchNameEnvironment.java:581]: > Its not a small change.

In that case we should separate the change with the actual fix from the patch that changes the usage of "MODULE_PATH" container.

Regarding https://github.com/trancexpress/eclipse.jdt.core/commit/404f642d03eb2f1e32013552867a24d595db176a : I'm not sure I can follow. The "MODULE_PATH" seem to be a shortcut to include all required modules of a project module, the patch I see in the branch seem to simply add the "module" attribute to all classpath entries, not the required modules. `org.eclipse.jdt.internal.core.ModulePathContainer.getClasspathEntries()`.

So I don't understand what exactly breaks with the original patch in "MODULE_PATH" and how the second patch supposed to fix that.

Would be nice first to explain the fail and discuss the right solution after that?

Comment (trancexpress) [reply org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/JavaSearchNameEnvironment.java:581]: Opened #717 to check if/why we need this entry in tests.

> Would be nice first to explain the fail and discuss the right solution after that?

The new fails are due to expecting the `module` attribute for classpath entries on the module path, in order to add module information during the search. The tests that fail require having classpath entries on the module path, but don't explicitly make those entries modular. This worked, since the search environment adds module infos for every classpath entry (assuming JLS > 9). Now it no longer works.

**Comment by trancexpress** [comment]
Some fails in `JavaSearchBugs9Tests` remain, I'm not sure how to fix those. I think we have a general problem with source classpath entries not having the module attribute. Since they don't have the attribute, we would not add the module info... But its possible that the project has a module path and everything is on the module path. I don't know what JDT does to ensure the sources are also on the module path for the compile...

**Comment by trancexpress** [comment]
> So we have to figure out what is the best way to fix them.

That is simple, we can adjust the change to not exclude source classpath entries.

**tl;dr** I don't see a reliable way to see if a source folder of a project is on the module path or on the classpath, during the search

However the new test fails. We can of course change the new test to use a library, same as the original problem we ran into. Of course the case where actual referenced projects are on the classpath (and not on the module path) will have the same problem.

Unfortunately the package fragment roots that pass through the changed code are the source folders of each individual project. I.e. if I have this structure:

```
project1 -> depends on project2
    src
project2
    src
```

What I see is `src` from project1` and `src` from `project2`. Neither of which has the `module` attribute. I.e. I see 2 classpath entries of 2 different projects:

```
<classpathentry kind="src" path="src"/>
```
```
<classpathentry kind="src" path="src"/>
```

If instead the environment was created with `src` from project` and `src` from the dependency to `src` from `project2`, there would be something to work with. Since the latter has the `module` attribute if its on the module path; we can rely on its presence or absence:

```
	<classpathentry kind="src" path="src"/>
	<classpathentry kind="src" path="/project2"/>
```

I'm not sure what we can do about this... In theory, we can examine the projects referencing the current project, checking their whether the reference is modular or not... But that assumes there is 1 "main" project with which to judge.

**Comment by trancexpress** [comment]
Now the new test fails. But we can rewrite it to use jars instead of project references and open a new issue to try to fix the case of referenced projects. I really don't know what to do for that case. Without having a "main" project from which the search is initiated there is nothing to be done I think. I'll see if the search environment knows from where it was initiated...

We might want to define `JavaSearchNameEnvironment.isModularProject(IJavaProject)` somewhere where `JavaRuntime` in JDT debug can use it.

**Comment by trancexpress** [comment]
@iloveeclipse see my last 2 comments, I think we have to discuss how to proceed here. I was naive to think the `module` attribute is always present. It isn't for source folders of projects containing those source folders. This complicates things a lot.

We can focus on fixing the initial case, problems in libraries (for those the `module` attribute is set).

**Comment by iloveeclipse** [comment]
`JavaSearchNameEnvironment.addModuleClassPathInfo(PackageFragmentRoot, IModuleDescription, ClasspathLocation)` asks for module descriptor and if that is present on a **library** entry, we should check "module" attribute, not before. So new logik should go a bit later => no problems with source folders anymore.

**Comment by trancexpress** [comment]
> `JavaSearchNameEnvironment.addModuleClassPathInfo(PackageFragmentRoot, IModuleDescription, ClasspathLocation)` asks for module descriptor and if that is present on a **library** entry, we should check "module" attribute, not before. So new logik should go a bit later => no problems with source folders anymore.

The new test will still fail. It shows the problem with source folders, not libraries. I.e. as said above, we'll fix only a part of the problem.

**Comment by trancexpress** [comment]
I've adjusted the new test to use libraries and not referenced projects.

> `JavaSearchNameEnvironment.addModuleClassPathInfo(PackageFragmentRoot, IModuleDescription, ClasspathLocation)` asks for module descriptor and if that is present on a **library** entry, we should check "module" attribute, not before. So new logik should go a bit later

I didn't do this since there are many `addModuleClassPathInfo()` methods doing different things based on module information available here and there. From my POV better to add code in one place than all the places in the different `addModuleClassPathInfo()` methods.

**Comment by trancexpress** [comment]
I've created an extra branch, to showcase what I meant with my previous comments:

https://github.com/trancexpress/eclipse.jdt.core/commit/4eaa334e6a5104f5f1595e22da8ab01240896aae

If we want to fix also the case for referenced projects, we can likely refine that commit. If we pass the project from which the search was started properly (currently its the first element in the passed `ArrayList` of projects, I'm using this to showcase the bug fix), maybe its worthwhile to make that change as well.

It would be a bigger change though. And as I've commented, I'm not 100% sure how valid this is. For the example in the new test its valid to do this, since the type used in the search is defined in the project. But if the search is initiated for a type found in a library, I'm not sure its OK to do this... Since the library can be on the classpath of any number of projects, none of which would be the "main" one then...

**Comment by trancexpress** [comment]
> I've created an extra branch, to showcase what I meant with my previous comments:
> 
> [trancexpress@4eaa334](https://github.com/trancexpress/eclipse.jdt.core/commit/4eaa334e6a5104f5f1595e22da8ab01240896aae)
> 
> If we want to fix also the case for referenced projects, we can likely refine that commit. If we pass the project from which the search was started properly (currently its the first element in the passed `ArrayList` of projects, I'm using this to showcase the bug fix), maybe its worthwhile to make that change as well.

I guess another option would be to check which projects are on the classpath and which are on the module path, in `MatchLocator`. Then pass this information to `JavaSearchNameEnvironment`. Then when we see a source entry from a project, we would know if its on the module path.

We can maybe do this only for searches where the type has a source compilation unit. That way we are sure the search really is about the project where the search started.

**Comment by rgrunber** [comment]
Is there any way this change might be related to type signature resolution of local variable type signatures ?

We recently started seeing test failures in a project that uses the JDT Core API, https://ci.eclipse.org/lsp4mp/job/lsp4mp/job/master/211/testReport/ . If I import jdt.core into my workspace (as opposed to from the target platform), the test failures begin happening at this commit.

If I take a `LocalVariable` element (eg. whose type is `String`) and calls `getTypeSignature()`. The result used to be `Ljava.lang.String` whereas now it is `LString`. I can't find any example/references to `LString` as a valid signature. It should either be `QString`, or `Ljava.lang.String`.

**Comment by iloveeclipse** [comment]
There was a regression, see https://github.com/eclipse-jdt/eclipse.jdt.core/issues/740, and the fix was just merged, see https://github.com/eclipse-jdt/eclipse.jdt.core/pull/746.

So if that was the problematic commit, please check next I build.

**Comment by iloveeclipse** [comment]
@rgrunber : I20230214-1800 build should contain the fix. Please test.

**Comment by iloveeclipse** [comment]
Thanks Roland!
